onebeyond/license-checker
# refactor!: license arguments spdx compliant

### Main changes

- :new: (feat)
- :wrench: (chore)
- :memo: (docs)
- :arrows_counterclockwise: (refactor)
- :microscope: (test)
- :bangbang: (breaking change)

This PR is a **BREAKING CHANGE**. The main goal of this PR is to keep the current functionality but making the scanning licenses process SPDX compliant. For that, I've done the following:

- Remove the utility functions related to the old approach for the matching process. We were using string or RegEx's passed to `failOn` to check if any of the packages' licenses matched the value. As we no longer need that, I've removed all the related functions and their tests
- Remove the `spdx-validator` component introduced in the previous PR. The only reason for this component to exist was the wrapper around the `spdx-expression-validate` to throw an error when the license wasn't SPDX compliant. This was used to check the value of `checkLicense` option. I've moved this code from the `index.js` to `runner.js`, and control the error there. So there is no need to keep the `spdx-validator` component anymore, we import the SPDX packages on the `runner.js` and use them
- `failOn` stays as a list and `generateOutputOn` has been changed to a list for consistency. I've updated its description and the type on the `args.js` file (the `yargs` setup), and the documentation.
- Remove the args' middleware. The old functions were used to verify if the RegEx values passed to `failOn` were valid, and to parse the `generateOutput` licenses. These were meant to be passed as a comma separated list of licenses, which no longer applies.
- Refactor the `runner` logic to use `spdx-validator` and `spdx-expression-validate` packages.
  - As mentioned before, the `runner` now controls the `checkLicenses` flow. If a value is passed, it will check it and finish without triggering the scanning process
  - The list of licenses passed to `failOn` and `generateOutputOn` options, if provided, must be a SPDX compliant. Otherwise, an error will be thrown. If all are correct, they are concatenated using the `OR` operand. The result of this operation is used for the matching process. So if executed with `--failOn MIT GPL-1.0 0BSD`, the SPDX expression to check the package licenses will be `MIT OR GPL-1.0 OR 0BSD`
  - Packages' licenses must also be SPDX. So if any of the `package.json` entries has a non-compliant license, it won't be taken into account for the scanning process. A warning message will be printed to inform what packages are in this situation. We could potentially include an option to fail if that's the case.

To sum up, the main functionality stays the same. However, the checking process will make use of the SPDX packages. This will provide support for the `AND`, `OR` and `WITH` syntax during the matching process, which previously wasn't possible. 

### Additional notes

README has been updated accordingly. We've found an issue in spdx-satisfies package. An explanation has been added to this document as well.
I have had some errors with eslint (incompatible packages). So I've reinstalled all the `eslint` dependencies and replaced the `browser` option by `node` on the `.eslintrc.json`

### Context

- :ticket: Closes / Relates #53
